flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkMetaDataTable.java

```diff
@@ -0,0 +1,713 @@
+    configuration.set(FlinkConfigOptions.TABLE_EXEC_ICEBERG_USE_FLIP27_SOURCE, true);
+    return super.getTableEnv();
+  }
+
+  @Before
+  public void before() {
+    super.before();
+    sql("CREATE DATABASE %s", flinkDatabase);
+    sql("USE CATALOG %s", catalogName);
+    sql("USE %s", DATABASE);
+  }
+
+  @Override
+  @After
+  public void clean() {
+    sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, TABLE_NAME);
+    sql("DROP DATABASE IF EXISTS %s", flinkDatabase);
+    super.clean();
+  }
+
+  @Test
+  public void testSnapshots() {
+    sql(
+        "CREATE TABLE %s (id INT, data VARCHAR,d DOUBLE) WITH ('format-version'='2', 'write.format.default'='%s')",
+        TABLE_NAME, format.name());
+    sql(
+        "INSERT INTO %s VALUES (1,'iceberg',10),(2,'b',20),(3,CAST(NULL AS VARCHAR),30)",
+        TABLE_NAME);
+    String sql = String.format("SELECT * FROM %s$snapshots ", TABLE_NAME);
+    List<Row> result = sql(sql);
+
+    Table table = validationCatalog.loadTable(TableIdentifier.of(icebergNamespace, TABLE_NAME));
+    Iterator<Snapshot> snapshots = table.snapshots().iterator();
+    for (Row row : result) {
+      Snapshot next = snapshots.next();
+      Assert.assertEquals(
+          "Should have expected timestamp",
+          ((Instant) row.getField(0)).toEpochMilli(),
+          next.timestampMillis());
+      Assert.assertEquals("Should have expected snapshot id", row.getField(1), next.snapshotId());
+      Assert.assertEquals("Should have expected parent id", row.getField(2), next.parentId());
+      Assert.assertEquals("Should have expected operation", row.getField(3), next.operation());
+      Assert.assertEquals(
+          "Should have expected manifest list location",
+          row.getField(4),
+          next.manifestListLocation());
+      Assert.assertEquals("Should have expected summary", row.getField(5), next.summary());
+    }
+  }
+
+  @Test
+  public void testHistory() {
+    sql(
+        "CREATE TABLE %s (id INT, data VARCHAR,d DOUBLE) WITH ('format-version'='2', 'write.format.default'='%s')",
+        TABLE_NAME, format.name());
+    sql(
+        "INSERT INTO %s VALUES (1,'iceberg',10),(2,'b',20),(3,CAST(NULL AS VARCHAR),30)",
+        TABLE_NAME);
+    String sql = String.format("SELECT * FROM %s$history ", TABLE_NAME);
+    List<Row> result = sql(sql);
+
+    Table table = validationCatalog.loadTable(TableIdentifier.of(icebergNamespace, TABLE_NAME));
+    Iterator<Snapshot> snapshots = table.snapshots().iterator();
+    for (Row row : result) {
+      Snapshot next = snapshots.next();
+      Assert.assertEquals(
+          "Should have expected made_current_at",
+          ((Instant) row.getField(0)).toEpochMilli(),
+          next.timestampMillis());
+      Assert.assertEquals("Should have expected snapshot id", row.getField(1), next.snapshotId());
+      Assert.assertEquals("Should have expected parent id", row.getField(2), next.parentId());
+      Assert.assertEquals(
```

Comment:
this is where having multiple snapshots can make the verification more comprehensive. E.g., it wasn't immediately clear to me what `is_current_ancestor` means. is it only for the table's current snapshot?

Reply:
Multiple snapshots have been added.